accel/opencl/ops_opencl.py

```diff
@@ -22,14 +22,45 @@ def load(x):
 MATMUL_SRC = load(pathlib.Path(__file__).resolve().parent.parent.parent / 'accel/opencl/matmul.cl')
 
 class CLImage:
+  MAX_HW = 16_384
   fmt = cl.ImageFormat(cl.channel_order.RGBA, cl.channel_type.HALF_FLOAT if FLOAT16 else cl.channel_type.FLOAT)
 
   def __init__(self, shape):
-    self.cl = cl.Image(CL().cl_ctx, cl.mem_flags.READ_WRITE, CLImage.fmt, shape=shape)
+    self.shape = shape
+    self.n_tile = int(np.ceil(max(shape) / CLImage.MAX_HW).item())
+    # if n_tile > 1, we can't fit the image into a CL image at native size,
+    # and need to internally store it as a set of disjoint tiles
+    if self.n_tile * min(shape) > CLImage.MAX_HW:
+      raise Exception(f"shape {shape} exceeds Metal image limits, even after tiling")
+    if shape[0] >= shape[1]:
+      # wider than it is tall; extra tiles overflow on y
+      self.tile_axis, tiled_width, tiled_height = 1, min(shape[0], CLImage.MAX_HW), self.n_tile * shape[1]
+    else:
+      # taller than it is wide; extra tiles overflow on x
+      self.tile_axis, tiled_width, tiled_height = 0, self.n_tile * shape[0], min(shape[1], CLImage.MAX_HW)
+    if DEBUG >= 1:
+      print(f"To create image with shape {shape}, made a tiled image with shape {tiled_width, tiled_height} and n_tile {self.n_tile} and tile axis {self.tile_axis}")
+      print(f"   to get coords for l, we sample {self.pos_to_sample_pos()}")
+    self.cl = cl.Image(CL().cl_ctx, cl.mem_flags.READ_WRITE, CLImage.fmt, shape=(tiled_width, tiled_height))
     CL.mem_used += self.cl.row_pitch * self.cl.height
 
+  def pos_to_sample_pos(self, l="l", check_bounds=True):
+    if self.n_tile == 1:
+      # happy path where no indexing ops are needed
+      return l
+    # sad tiled path
+    if self.tile_axis == 1:
+      sample_pos = f"((int2)({l}.x % {CLImage.MAX_HW}, ({l}.x / {CLImage.MAX_HW}) * {self.shape[1]} + {l}.y))"
+    else:
+      sample_pos = f"((int2)(({l}.y / {CLImage.MAX_HW}) * {self.shape[0]} + {l}.x, {l}.y % {CLImage.MAX_HW}))"
+    if check_bounds:
+      in_bounds = f"((0 <= {l}.x) && ({l}.x < {self.shape[0]}) && (0 <= {l}.y) && ({l}.y < {self.shape[1]}))"
```

Comment:
Do you have to check the 0? If it's less than 0, it'll already always be out of bounds, right?

Reply:
Fixed, to skip the lower-bound check on whichever axis doesn't need it.